### PR TITLE
[stm] Make the main interfaces functional.

### PR DESCRIPTION
**Pull request by ejgallego**
We make the main STM interfaces functional, removing most state from
`stm.ml`.

In particular, error handling has changed quite a lot, as we must not
restore the document. Further improvements are possible in this area.

With regards to the worker interface, we include the document in the
functional task, this should be OK for now but in the future we may
need to use more refined rules.

After some discussion with @gares we have also removed the "keep
states" optimization in the proof worker. It should be introduced in
the future once we make the TaskQueue API aware of worker state, but
IMHO TaskQueue improvements should go in a separate commit.

Be should be careful to check if there is any performance impact thou.

**Note: ** This is a preliminary PR as to communicate some work in this area.

### CODE REVIEW

File: stm/stm.ml [hunk 1]
```diff
@@ -1168,6 +1142,8 @@ end = struct (* {{{ *)
 end (* }}} *)
 
 let hints = ref Aux_file.empty_aux_file
+(* let set_compilation_hints ~doc file = *)
+(* XXX *)
```
Comment (gares) [inline stm/stm.ml]: Are you saying that hints are still imperative?

File: stm/stm.ml [hunk 2]
```diff
@@ -1339,7 +1307,6 @@ end = struct (* {{{ *)
   type response =
     | RespBuiltProof of Proof_global.closed_proof_output * float
     | RespError of error
-    | RespStates of (Stateid.t * State.partial_state) list
```
Comment (gares) [inline stm/stm.ml]: The entire logic for sending back a state is gone. Are you saying that with this patch, if one edits a broken proof, all steps frmo the beginning of the proof to the error are recomputed?

**Review by gares — CHANGES_REQUESTED** [review]
Unless I'm mistaken, the purely functional `doc` makes it impossible to fill in the cache of computed states asynchronously. I don't see how a cache can be purely functional in program with threads: who passes the cache to the current thread? who passes it back to the next one?

One could get rid of threads, use cooperative code as in lwt and have the scheduler pass the cache.  But this is not in this patch.

This patch is nice, but it is too drastic. The cache shall stay imperative.  It is a cache, so if it gets magically filled in, then all references to a doc see it, it is just fine (semantically I mean).

The change I request to put the cache back, while the functional API is a clear cleanup and shall stay.

File: stm/stm.ml (hunk 2 above)
Comment (ejgallego) [reply stm/stm.ml]: I don't know @gares , I did that as I understood from our conversation that this was a non-very important optimization. I can resurrect that part without a problem, but will have to tweak the interface a bit so workers declare local persistent state.

**Comment by ejgallego** [comment]
> Unless I'm mistaken, the purely functional doc makes it impossible to fill in the cache of computed states asynchronously. I don't see how a cache can be purely functional in program with threads: who passes the cache to the current thread? who passes it back to the next one?
> 
> One could get rid of threads, use cooperative code as in lwt and have the scheduler pass the cache. But this is not in this patch.
> 
> This patch is nice, but it is too drastic. The cache shall stay imperative. It is a cache, so if it gets magically filled in, then all references to a doc see it, it is just fine (semantically I mean).
> 
> The change I request to put the cache back, while the functional API is a clear cleanup and shall stay.

Well, I tried to remove the possibility to fill states asynchronously following our conversion maybe I misunderstood.

Indeed if we are going to keep the filling of states then you raise a fair point but I digress a bit on how to handle the cache. I agree that it should use imperative futures (that get resolved) but in essence the cache (and the reference to this futures) should be handled functionally from the point of view of the STM code, so maybe one day we can implement a monadic pure API for async work. This is the intention of what happens in the patch with proofs.

On the other hand, I think that this patch needs a lot more work. In fact, I am not sure that modifying the STM is the best way to achieve the goals I do have in mind, as opposed to hooking a new proof manager just after vernac. I guess I'll have to think a bit...

**Comment by ejgallego** [comment]
As to be more concrete, the cache could indeed be:

```
type st_cache = (VernacState.t Future.t) StateId.Map
```

On the other hand, we have several asynchronous processing mechanisms. Umm, on one side we get feedback, on the other we have the futures, and on the other we have these task completion things...

Umm... I see many possibilities, but it is hard to organize thoughts into a plan.

File: stm/stm.ml (hunk 2 above)
Comment (gares) [reply stm/stm.ml]: What I mean was "always send back state(s) as part of the response to the other message"

Comment (ejgallego) [reply stm/stm.ml]: Ah! Understood, thanks.

**Comment by gares** [comment]
My impression is that such cache better be part of the document's nodes. E.g. if you remove a node, also its associated state (if present) is collected.  If separate the document from the cache, you have keep them in sync, or use weak tables, or your cache will leak.

Note that it is not really a cache: once an item is in there, as long as it is reference, you should not remove it (you can, and it won't change the semantics of the document, but it is not what the user expects IMHO). So being part of the document faithfully represent this property.

**Comment by gares** [comment]
Anyway, I'd like this PR to be merged, so what I suggest is that you restore the 1 mutable bit of the document (the state option) and restore the code sending back states. All the rest of the code is an improvement, and putting it back this little imperative bit is not going to change much your API

**Comment by ejgallego** [comment]
Maybe the most pressing reason to separate the document from the cache is to avoid having the "ghost" nodes, which are used for side-effect replay. Indeed we should have `span_id` and `exec_id` with several `exec_id` allowing to refer to the same `span_id`.

In fact, getting rid of the ghost nodes is IMVHO the top blocker in order to advance the IDE protocol.

**Comment by gares** [comment]
Hum, look like an XY problem then. There is little relation between the ghost node and separating the cache from the document. 

What you want is to send 2 state_id in response to an add, and have the UI aggregate feedback from both ids. One of them (the first one for example) would be the one the UI can use to point to that particular "state change" (for edit_at).

Indeed there is a parallel with Isabelle's multiple exec id for a single span id.

Still this PR is (but for the comment in my review) making the code nicer, so please fix it, let's merge it, and then let's try to fix the protocol/document to make ghost nodes less painful (in a way or another).

**Comment by ejgallego** [comment]
> What you want is to send 2 state_id in response to an add, and have the UI aggregate feedback from both ids.

No, I don't want this. My life as UI implementor would be hell. To an `add` I want no response. It is the UI who controls the document, not Coq.

"Ghost nodes" have to go. They are an implementation detail that shouldn't leak to the UI and much less we should make the UI handle that.

> There is little relation between the ghost node and separating the cache from the document.

The is full relation. Ghost nodes are an artifact of execution, as well as the cache is a property associated to the execution, not to the structure of the document.

My view is very clear here, Coq has to right to interfere with the document, the UI has full control of it.

Indeed Isabelle clearly distinguished about document and execution, but coq does not and hence the current problems for writing an IDE. But this is in a sense bad news, as the assumption that the document reflect the scheduling is unfortunately at the heart of the current design :S

**Comment by ejgallego** [comment]
Let me extend a bit more on the semantics of `Add`: synchronous calls are worse than poison on the UI side.

In most UI paradigms, you cannot just block the user from doing an action before a synchronous call has completed.

We **must** eliminate all sync UI calls in Coq.

**Comment by gares** [comment]
I see an update, but the states are still not sent-back/assigned

**Comment by ejgallego** [comment]
I'm much afraid that this needs major rework IMHO.

**Comment by maximedenes** [comment]
@ejgallego Should this PR be closed? I see you moved it to "obsolete".

**Comment by ejgallego** [comment]
As @gares mentioned with the current imperative update from workers this approach is going to be defeated every time; I am trying out different things and I have reached the point where this can be safely deprecated.

**Pull request closed without merging**